feat(gateway): GCP Postgres Auth (#3679)
Kong/developer.konghq.com: app/gateway/gcp-postgres-authentication.md

With authentication enabled, you don't need a password to connect to a database instance. Instead, you use a temporary authentication token. Because GCP manages the authentication externally, the database doesn't store user credentials. If you're using Google Cloud Postgres for {{site.base_gateway}}'s database, you can enable authentication on your running cluster. This eliminates the need to store user credentials on both the {{site.base_gateway}} (`pg_password`) and Google Cloud Postgres sides.

## GCP authentication limitations

GCP authentication has some limitations. Go through each one before you use this feature in your production environment:

* This feature cannot be used together with databases from other cloud providers, such as [AWS RDS](/gateway/amazon-rds-authentication-with-aws-iam/). These auth providers are mutually exclusive. 
* When `pg_gcp_auth` is enabled, the `pg_password` won't be used. You can't use both methods at the same time.

> * Any incorrect configuration on the GCP side will result in a failure in initializing the database connection. For example, an improperly configured managed identity or not creating a corresponding role inside GCP Postgres.

Reviewer: ```suggestion
* Any incorrect configuration on the GCP side will result in a failure in initializing the database connection, such as an improperly configured managed identity or a missing role inside GCP Postgres.
```

For additional recommendations and limitations, see the [IAM authentication restrictions](https://docs.cloud.google.com/sql/docs/postgres/iam-authentication#restrictions) in the Google Cloud documentation.

## Enabling GCP authentication

> You can enable GCP authentication through an environment variable or the {{site.base_gateway}} configuration file. It supports both read-only and read-write modes, or you can enable it in read-only mode only.

Reviewer: ```suggestion
You can enable GCP authentication through an environment variable or the {{site.base_gateway}} configuration file. You can enable it for both read-write and read-only modes, or for read-only mode only.
```

KONG_PG_RO_GCP_AUTH=on
```

Then, set the following, replacing placeholders with your values:
```bash
KONG_PG_USER='username@project-name.iam'        # Postgres user.
KONG_PG_DATABASE='kong'                         # The database name to connect to.
KONG_PG_HOST='35.200.xx.yy'                     # Host of the Postgres server.
KONG_PG_PORT='5432'                             # Port of the Postgres server.
KONG_PG_GCP_SERVICE_ACCOUNT_JSON='{"type":"service_account","project_id":"example-project-294816","private_key_id":"a7b3c9d2e8f1g4h6i5j7k9m2n4p6q8r1","private_key":"-----BEGIN PRIVATE KEY-----\nMIIEvQIBADANB…………………………..…5xX5yY5zA==\n-----END PRIVATE KEY-----\n","client_email":"example-sa@example-project-294816.iam.gserviceaccount.com","client_id":"103847562938475629384","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://oauth2.googleapis.com/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_x509_cert_url":"https://www.googleapis.com/robot/v1/metadata/x509/example-sa%40example-project-294816.iam.gserviceaccount.com","universe_domain":"googleapis.com"}'
```

{:.info}

> > **Workload Identity only:** If you're using the Workload Identity to authenticate, you don't need to configure the `KONG_PG_GCP_SERVICE_ACCOUNT_JSON`. {{site.base_gateway}}'s GCP authentication feature is designed to automatically fall back to the Workload Identity mechanism if the GCP service account JSON key isn't found in the configuration.

Reviewer: ```suggestion
> **Workload Identity only:** If you're using Workload Identity to authenticate, you don't need to configure the `KONG_PG_GCP_SERVICE_ACCOUNT_JSON`. {{site.base_gateway}}'s GCP authentication feature is designed to automatically fall back to the Workload Identity mechanism if the GCP service account JSON key isn't found in the configuration.
```
